onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,512 @@
+  std::vector<int64_t> ngram_indexes_;
+  std::vector<float> weights_;
+
+  std::vector<std::string> pool_strings_;
+  StringPoolSet str_set_;
+  IntegerPoolSet int_set_;
+  size_t output_size_ = 0;
+
+  MLDataType int32_dt_;
+  MLDataType int64_dt_;
+  MLDataType string_dt_;
+  Impl() {
+    int32_dt_ = DataTypeImpl::GetType<int32_t>();
+    int64_dt_ = DataTypeImpl::GetType<int64_t>();
+    string_dt_ = DataTypeImpl::GetType<std::string>();
+  }
+
+  template <typename T>
+  auto PoolEnd() const;
+
+  template <typename T>
+  auto Find(const ngram_details::NGramItem<T>&) const;
+
+  void IncrementCount(size_t ngram_id, std::vector<uint32_t>& frequencies) const {
+    assert(ngram_id < ngram_indexes_.size());
+    auto output_idx = ngram_indexes_[ngram_id];
+    ORT_ENFORCE(output_idx >= 0, "ngram_indxes has a negative index");
+    assert(static_cast<size_t>(output_idx) < frequencies.size());
+    ++frequencies[output_idx];
+  }
+};
+
+template <>
+inline auto Ngram::Impl::PoolEnd<int64_t>() const {
+  return int_set_.cend();
+}
+
+template <>
+inline auto Ngram::Impl::PoolEnd<int32_t>() const {
+  return PoolEnd<int64_t>();
+}
+
+template <>
+inline auto Ngram::Impl::PoolEnd<std::string>() const {
+  return str_set_.cend();
+}
+
+template <>
+inline auto Ngram::Impl::Find<int64_t>(const NGramItem<int64_t>& i) const {
+  return int_set_.find(i);
+}
+
+template <>
+inline auto Ngram::Impl::Find<int32_t>(const NGramItem<int32_t>& i) const {
+  return int_set_.find(i);
+}
+
+template <>
+inline auto Ngram::Impl::Find<std::string>(const NGramItem<std::string>& i) const {
+  return str_set_.find(i);
+}
+
+Ngram::Ngram(const OpKernelInfo& info) : OpKernel(info), impl_(new Impl) {
+  std::string mode;
+  Status status = info.GetAttr("mode", &mode);
+  ORT_ENFORCE(status.IsOK(), "mode is required");
+  if (mode == "TF") {
+    impl_->mode_ = kTF;
+  } else if (mode == "IDF") {
+    impl_->mode_ = kIDF;
+  } else if (mode == "TFIDF") {
+    impl_->mode_ = kTFIDF;
+  }
+  ORT_ENFORCE(impl_->mode_ != kNone, "Unrecognized mode");
+
+  status = info.GetAttr("M", &impl_->M_);
+  ORT_ENFORCE(status.IsOK() && impl_->M_ > 0, "Positive Attr M is required");
+  status = info.GetAttr("N", &impl_->N_);
+  ORT_ENFORCE(status.IsOK() && impl_->N_ >= impl_->M_, "Positive M >= N is required");
+  status = info.GetAttr("S", &impl_->S_);
+  ORT_ENFORCE(status.IsOK() && impl_->N_ >= 0, "Non-negative number of skips S is required");
+
+  int64_t all = 0;
+  status = info.GetAttr("all", &all);
+  ORT_ENFORCE(status.IsOK(), "Attribute all is required");
+  impl_->all_ = (all != 0);
+
+  status = info.GetAttrs(std::string("ngram_counts"), impl_->ngram_counts_);
+  ORT_ENFORCE(status.IsOK() && !impl_->ngram_counts_.empty(), "Non-empty ngram_counts is required");
+  ORT_ENFORCE(size_t(impl_->M_) <= impl_->ngram_counts_.size(), "M must be inbounds of ngram_counts");
+  ORT_ENFORCE(size_t(impl_->N_) <= impl_->ngram_counts_.size(), "N must be inbounds of ngram_counts");
```

Comment:
It'd be nice if we can print out the actual values we got in those error messages.